Cairo fails to load Arial font on recent macOS version

See https://github.com/romgrk/node-gtk/pull/230#issuecomment-717858019

We need to find out why the status is `Cairo.Status.INVALID_STRING`. Maybe there's something which needs to be installed since the issue does not occur on travis.

> font.status() from const font = Cairo.FontFace.create('Arial', Cairo.FontSlant.OBLIQUE, Cairo.FontWeight.BOLD) is Cairo.Status.INVALID_STRING instead of Cairo.Status.SUCCESS. This issue is not present in travis.

I have these errors, one might be relevant?

1 warning generated.
  CXX(target) Release/obj.target/node_gtk/src/modules/cairo/scaled-font.o
In file included from ../src/modules/cairo/scaled-font.cc:4:
In file included from ../src/modules/cairo/../../gi.h:7:
In file included from ../node_modules/nan/nan.h:2884:
../node_modules/nan/nan_typedarray_contents.h:34:43: warning: 'GetContents' is deprecated: Use GetBackingStore. See http://crbug.com/v8/9908. [-Wdeprecated-declarations]
      data   = static_cast<char*>(buffer->GetContents().Data()) + byte_offset;
                                          ^
/Users/webreflection/Library/Caches/node-gyp/15.0.1/include/node/v8.h:5310:3: note: 'GetContents' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use GetBackingStore. See http://crbug.com/v8/9908.")
  ^
/Users/webreflection/Library/Caches/node-gyp/15.0.1/include/node/v8config.h:402:39: note: expanded from macro 'V8_DEPRECATE_SOON'
# define V8_DEPRECATE_SOON(message) [[deprecated(message)]]
                                      ^
../src/modules/cairo/scaled-font.cc:251:18: warning: unused variable 'result' [-Wunused-variable]
  cairo_status_t result = cairo_scaled_font_text_to_glyphs (scaled_font,

--- a/src/modules/cairo/font-face.cc
+++ b/src/modules/cairo/font-face.cc
@@ -424,7 +424,7 @@ NAN_METHOD(FontFace::getReferenceCount) {
 
 NAN_METHOD(FontFace::create) {
   // in-arguments
-  auto family__value = info[0].As<String>(); auto family = *Nan::Utf8String(family__value);
+  Nan::Utf8String family__value(info[0].As<String>()); auto family = *family__value;
   auto slant = (cairo_font_slant_t) Nan::To<int64_t>(info[1].As<Number>()).ToChecked();
   auto weight = (cairo_font_weight_t) Nan::To<int64_t>(info[2].As<Number>()).ToChecked();
 

--- a/src/modules/cairo/font-face.cc
+++ b/src/modules/cairo/font-face.cc
@@ -424,7 +424,7 @@ NAN_METHOD(FontFace::getReferenceCount) {
 
 NAN_METHOD(FontFace::create) {
   // in-arguments
-  auto family = *Nan::Utf8String (info[0].As<String>());
+  auto family__value = info[0].As<String>(); auto family = *Nan::Utf8String(family__value);
   auto slant = (cairo_font_slant_t) Nan::To<int64_t>(info[1].As<Number>()).ToChecked();
   auto weight = (cairo_font_weight_t) Nan::To<int64_t>(info[2].As<Number>()).ToChecked();
 

--- a/src/modules/cairo/font-options.cc
+++ b/src/modules/cairo/font-options.cc
@@ -305,7 +305,7 @@ NAN_METHOD(FontOptions::setVariations) {
   auto options = Nan::ObjectWrap::Unwrap<FontOptions>(self)->_data;
 
   // in-arguments
-  auto variations = *Nan::Utf8String (info[0].As<String>());
+  auto variations__value = info[0].As<String>(); auto variations = *Nan::Utf8String(variations__value);
 
   // function call
   cairo_font_options_set_variations (options, variations);

--- a/src/modules/cairo/font-options.cc
+++ b/src/modules/cairo/font-options.cc
@@ -305,7 +305,7 @@ NAN_METHOD(FontOptions::setVariations) {
   auto options = Nan::ObjectWrap::Unwrap<FontOptions>(self)->_data;
 
   // in-arguments
-  auto variations__value = info[0].As<String>(); auto variations = *Nan::Utf8String(variations__value);
+  Nan::Utf8String variations__value(info[0].As<String>()); auto variations = *variations__value;
 
   // function call
   cairo_font_options_set_variations (options, variations);

--- a/src/modules/cairo/generator.js
+++ b/src/modules/cairo/generator.js
@@ -415,7 +415,7 @@ function getInArgumentSource(p, n) {
     return `auto ${p.name} = Nan::To<int64_t>(info[${n}].As<Number>()).ToChecked();`
 
   if (typeName === 'const char *')
-    return `auto ${p.name}__value = info[${n}].As<String>(); auto ${p.name} = *Nan::Utf8String(${p.name}__value);`
+    return `Nan::Utf8String ${p.name}__value(info[${n}].As<String>()); auto ${p.name} = *${p.name}__value;`
 
   if (baseName in ENUM_TYPE)
     return `auto ${p.name} = (${typeName}) Nan::To<${ENUM_TYPE[typeName]}>(info[${n}].As<Number>()).ToChecked();`

--- a/src/modules/cairo/generator.js
+++ b/src/modules/cairo/generator.js
@@ -415,7 +415,7 @@ function getInArgumentSource(p, n) {
     return `auto ${p.name} = Nan::To<int64_t>(info[${n}].As<Number>()).ToChecked();`
 
   if (typeName === 'const char *')
-    return `auto ${p.name} = *Nan::Utf8String (info[${n}].As<String>());`
+    return `auto ${p.name}__value = info[${n}].As<String>(); auto ${p.name} = *Nan::Utf8String(${p.name}__value);`
 
   if (baseName in ENUM_TYPE)
     return `auto ${p.name} = (${typeName}) Nan::To<${ENUM_TYPE[typeName]}>(info[${n}].As<Number>()).ToChecked();`

--- a/src/modules/cairo/scaled-font.cc
+++ b/src/modules/cairo/scaled-font.cc
@@ -191,7 +191,7 @@ NAN_METHOD(ScaledFont::textExtents) {
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
 
   // in-arguments
-  auto utf8 = *Nan::Utf8String (info[0].As<String>());
+  auto utf8__value = info[0].As<String>(); auto utf8 = *Nan::Utf8String(utf8__value);
 
   // out-arguments
   auto extents = Nan::NewInstance(
@@ -212,23 +212,23 @@ NAN_METHOD(ScaledFont::glyphExtents) {
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
 
   // in-arguments
-  auto glyphs = Nan::ObjectWrap::Unwrap<Glyph>(info[0].As<Object>());
+  auto glyphs = Nan::ObjectWrap::Unwrap<Glyph>(info[0].As<Object>())->_data;
+  auto num_glyphs = Nan::To<int64_t>(info[1].As<Number>()).ToChecked();
 
   // out-arguments
   auto extents = Nan::NewInstance(
-          Nan::New<Function>(TextExtents::constructor),
-          0,
-          NULL).ToLocalChecked();
+            Nan::New<Function>(TextExtents::constructor),
+            0,
+            NULL).ToLocalChecked();
 
   // function call
-  cairo_scaled_font_glyph_extents (scaled_font, glyphs->_data, glyphs->_length, Nan::ObjectWrap::Unwrap<TextExtents>(extents)->_data);
+  cairo_scaled_font_glyph_extents (scaled_font, glyphs, num_glyphs, Nan::ObjectWrap::Unwrap<TextExtents>(extents)->_data);
 
   // return
   Local<Value> returnValue = extents;
   info.GetReturnValue().Set(returnValue);
 }
 
-/* textToGlyphs(x, y, utf8, getClusters: boolean) */
 NAN_METHOD(ScaledFont::textToGlyphs) {
   auto self = info.This();
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
@@ -236,44 +236,32 @@ NAN_METHOD(ScaledFont::textToGlyphs) {
   // in-arguments
   auto x = Nan::To<double>(info[0].As<Number>()).ToChecked();
   auto y = Nan::To<double>(info[1].As<Number>()).ToChecked();
-  auto utf8 = *Nan::Utf8String (info[2].As<String>());
-  auto utf8_len = info[2].As<String>()->Length();
-  auto get_clusters = info.Length() == 4 ? Nan::To<bool>(info[3].As<v8::Boolean>()).ToChecked() : false;
+  auto utf8__value = info[2].As<String>(); auto utf8 = *Nan::Utf8String(utf8__value);
+  auto utf8_len = Nan::To<int64_t>(info[3].As<Number>()).ToChecked();
 
   // out-arguments
-  cairo_glyph_t *glyphs = NULL;
+  auto glyphs = Nan::NewInstance(
+            Nan::New<Function>(Glyph::constructor),
+            0,
+            NULL).ToLocalChecked();
   int num_glyphs = 0;
-  cairo_text_cluster_t *clusters = NULL;
+  auto clusters = Nan::NewInstance(
+            Nan::New<Function>(TextCluster::constructor),
+            0,
+            NULL).ToLocalChecked();
   int num_clusters = 0;
-  cairo_text_cluster_flags_t cluster_flags = (cairo_text_cluster_flags_t) 0;
+  cairo_text_cluster_flags_t *cluster_flags = NULL;
 
   // function call
-  cairo_status_t result = cairo_scaled_font_text_to_glyphs (scaled_font,
-      x,
-      y,
-      utf8,
-      utf8_len,
-      &glyphs,
-      &num_glyphs,
-      get_clusters ? &clusters : NULL,
-      get_clusters ? &num_clusters : NULL,
-      get_clusters ? &cluster_flags : NULL);
+  cairo_status_t result = cairo_scaled_font_text_to_glyphs (scaled_font, x, y, utf8, utf8_len, Nan::ObjectWrap::Unwrap<Glyph>(glyphs)->_data, &num_glyphs, Nan::ObjectWrap::Unwrap<TextCluster>(clusters)->_data, &num_clusters, &cluster_flags);
 
   // return
-  Local<Array> returnValue = Nan::New<Array> (get_clusters ? 2 : 1);
-  Local<Value> glyphsArgs[] = { Nan::New<External> (glyphs), Nan::New<Number> (num_glyphs) };
-  auto glyphsObject = Nan::NewInstance(
-      Nan::New<Function>(Glyph::constructor), 2, glyphsArgs).ToLocalChecked();
-  Nan::Set (returnValue, 0, glyphsObject);
-  if (get_clusters) {
-    Local<Value> clustersArgs[] = {
-          Nan::New<External> (clusters),
-          Nan::New<Number> (num_clusters),
-          Nan::New<Number> ((int64_t) cluster_flags) };
-    auto clustersObject = Nan::NewInstance(
-        Nan::New<Function>(TextCluster::constructor), 3, clustersArgs).ToLocalChecked();
-    Nan::Set (returnValue, 1, clustersObject);
-  }
+  Local<Object> returnValue = Nan::New<Object> ();
+  Nan::Set (returnValue, UTF8 ("glyphs"), Nan::New (glyphs));
+  Nan::Set (returnValue, UTF8 ("num_glyphs"), Nan::New (num_glyphs));
+  Nan::Set (returnValue, UTF8 ("clusters"), Nan::New (clusters));
+  Nan::Set (returnValue, UTF8 ("num_clusters"), Nan::New (num_clusters));
+  Nan::Set (returnValue, UTF8 ("cluster_flags"), Nan::New (cluster_flags));
   info.GetReturnValue().Set(returnValue);
 }
 

--- a/src/modules/cairo/scaled-font.cc
+++ b/src/modules/cairo/scaled-font.cc
@@ -191,7 +191,7 @@ NAN_METHOD(ScaledFont::textExtents) {
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
 
   // in-arguments
-  auto utf8__value = info[0].As<String>(); auto utf8 = *Nan::Utf8String(utf8__value);
+  Nan::Utf8String utf8__value(info[0].As<String>()); auto utf8 = *utf8__value;
 
   // out-arguments
   auto extents = Nan::NewInstance(
@@ -237,7 +237,7 @@ NAN_METHOD(ScaledFont::textToGlyphs) {
   // in-arguments
   auto x = Nan::To<double>(info[0].As<Number>()).ToChecked();
   auto y = Nan::To<double>(info[1].As<Number>()).ToChecked();
-  auto utf8__value = info[2].As<String>(); auto utf8 = *Nan::Utf8String(utf8__value);
+  Nan::Utf8String utf8__value(info[2].As<String>()); auto utf8 = *utf8__value;
   auto utf8_len = info[2].As<String>()->Length();
   auto get_clusters = info.Length() == 4 ? Nan::To<bool>(info[3].As<v8::Boolean>()).ToChecked() : false;
 

--- a/src/modules/cairo/scaled-font.cc
+++ b/src/modules/cairo/scaled-font.cc
@@ -207,28 +207,29 @@ NAN_METHOD(ScaledFont::textExtents) {
   info.GetReturnValue().Set(returnValue);
 }
 
+/* HAND-MODIFIED: glyphExtents(glyphs) */
 NAN_METHOD(ScaledFont::glyphExtents) {
   auto self = info.This();
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
 
   // in-arguments
-  auto glyphs = Nan::ObjectWrap::Unwrap<Glyph>(info[0].As<Object>())->_data;
-  auto num_glyphs = Nan::To<int64_t>(info[1].As<Number>()).ToChecked();
+  auto glyphs = Nan::ObjectWrap::Unwrap<Glyph>(info[0].As<Object>());
 
   // out-arguments
   auto extents = Nan::NewInstance(
-            Nan::New<Function>(TextExtents::constructor),
-            0,
-            NULL).ToLocalChecked();
+          Nan::New<Function>(TextExtents::constructor),
+          0,
+          NULL).ToLocalChecked();
 
   // function call
-  cairo_scaled_font_glyph_extents (scaled_font, glyphs, num_glyphs, Nan::ObjectWrap::Unwrap<TextExtents>(extents)->_data);
+  cairo_scaled_font_glyph_extents (scaled_font, glyphs->_data, glyphs->_length, Nan::ObjectWrap::Unwrap<TextExtents>(extents)->_data);
 
   // return
   Local<Value> returnValue = extents;
   info.GetReturnValue().Set(returnValue);
 }
 
+/* HAND-MODIFIED: textToGlyphs(x, y, utf8, getClusters: boolean) */
 NAN_METHOD(ScaledFont::textToGlyphs) {
   auto self = info.This();
   auto scaled_font = Nan::ObjectWrap::Unwrap<ScaledFont>(self)->_data;
@@ -237,31 +238,43 @@ NAN_METHOD(ScaledFont::textToGlyphs) {
   auto x = Nan::To<double>(info[0].As<Number>()).ToChecked();
   auto y = Nan::To<double>(info[1].As<Number>()).ToChecked();
   auto utf8__value = info[2].As<String>(); auto utf8 = *Nan::Utf8String(utf8__value);
-  auto utf8_len = Nan::To<int64_t>(info[3].As<Number>()).ToChecked();
+  auto utf8_len = info[2].As<String>()->Length();
+  auto get_clusters = info.Length() == 4 ? Nan::To<bool>(info[3].As<v8::Boolean>()).ToChecked() : false;
 
   // out-arguments
-  auto glyphs = Nan::NewInstance(
-            Nan::New<Function>(Glyph::constructor),
-            0,
-            NULL).ToLocalChecked();
+  cairo_glyph_t *glyphs = NULL;
   int num_glyphs = 0;
-  auto clusters = Nan::NewInstance(
-            Nan::New<Function>(TextCluster::constructor),
-            0,
-            NULL).ToLocalChecked();
+  cairo_text_cluster_t *clusters = NULL;
   int num_clusters = 0;
-  cairo_text_cluster_flags_t *cluster_flags = NULL;
+  cairo_text_cluster_flags_t cluster_flags = (cairo_text_cluster_flags_t) 0;
 
   // function call
-  cairo_status_t result = cairo_scaled_font_text_to_glyphs (scaled_font, x, y, utf8, utf8_len, Nan::ObjectWrap::Unwrap<Glyph>(glyphs)->_data, &num_glyphs, Nan::ObjectWrap::Unwrap<TextCluster>(clusters)->_data, &num_clusters, &cluster_flags);
+  cairo_status_t result = cairo_scaled_font_text_to_glyphs (scaled_font,
+      x,
+      y,
+      utf8,
+      utf8_len,
+      &glyphs,
+      &num_glyphs,
+      get_clusters ? &clusters : NULL,
+      get_clusters ? &num_clusters : NULL,
+      get_clusters ? &cluster_flags : NULL);
 
   // return
-  Local<Object> returnValue = Nan::New<Object> ();
-  Nan::Set (returnValue, UTF8 ("glyphs"), Nan::New (glyphs));
-  Nan::Set (returnValue, UTF8 ("num_glyphs"), Nan::New (num_glyphs));
-  Nan::Set (returnValue, UTF8 ("clusters"), Nan::New (clusters));
-  Nan::Set (returnValue, UTF8 ("num_clusters"), Nan::New (num_clusters));
-  Nan::Set (returnValue, UTF8 ("cluster_flags"), Nan::New (cluster_flags));
+  Local<Array> returnValue = Nan::New<Array> (get_clusters ? 2 : 1);
+  Local<Value> glyphsArgs[] = { Nan::New<External> (glyphs), Nan::New<Number> (num_glyphs) };
+  auto glyphsObject = Nan::NewInstance(
+      Nan::New<Function>(Glyph::constructor), 2, glyphsArgs).ToLocalChecked();
+  Nan::Set (returnValue, 0, glyphsObject);
+  if (get_clusters) {
+    Local<Value> clustersArgs[] = {
+          Nan::New<External> (clusters),
+          Nan::New<Number> (num_clusters),
+          Nan::New<Number> ((int64_t) cluster_flags) };
+    auto clustersObject = Nan::NewInstance(
+        Nan::New<Function>(TextCluster::constructor), 3, clustersArgs).ToLocalChecked();
+    Nan::Set (returnValue, 1, clustersObject);
+  }
   info.GetReturnValue().Set(returnValue);
 }
 

--- a/src/modules/cairo/surface.cc
+++ b/src/modules/cairo/surface.cc
@@ -342,7 +342,7 @@ NAN_METHOD(Surface::writeToPng) {
   auto surface = Nan::ObjectWrap::Unwrap<Surface>(self)->_data;
 
   // in-arguments
-  auto filename__value = info[0].As<String>(); auto filename = *Nan::Utf8String(filename__value);
+  Nan::Utf8String filename__value(info[0].As<String>()); auto filename = *filename__value;
 
   // function call
   cairo_status_t result = cairo_surface_write_to_png (surface, filename);
@@ -591,7 +591,7 @@ NAN_METHOD(Surface::supportsMimeType) {
   auto surface = Nan::ObjectWrap::Unwrap<Surface>(self)->_data;
 
   // in-arguments
-  auto mime_type__value = info[0].As<String>(); auto mime_type = *Nan::Utf8String(mime_type__value);
+  Nan::Utf8String mime_type__value(info[0].As<String>()); auto mime_type = *mime_type__value;
 
   // function call
   cairo_bool_t result = cairo_surface_supports_mime_type (surface, mime_type);
@@ -630,7 +630,7 @@ NAN_METHOD(Surface::unmapImage) {
 }
 NAN_METHOD(ImageSurface::createFromPng) {
   // in-arguments
-  auto filename__value = info[0].As<String>(); auto filename = *Nan::Utf8String(filename__value);
+  Nan::Utf8String filename__value(info[0].As<String>()); auto filename = *filename__value;
 
   // function call
   cairo_surface_t * result = cairo_image_surface_create_from_png (filename);

--- a/src/modules/cairo/surface.cc
+++ b/src/modules/cairo/surface.cc
@@ -342,7 +342,7 @@ NAN_METHOD(Surface::writeToPng) {
   auto surface = Nan::ObjectWrap::Unwrap<Surface>(self)->_data;
 
   // in-arguments
-  auto filename = *Nan::Utf8String (info[0].As<String>());
+  auto filename__value = info[0].As<String>(); auto filename = *Nan::Utf8String(filename__value);
 
   // function call
   cairo_status_t result = cairo_surface_write_to_png (surface, filename);
@@ -591,7 +591,7 @@ NAN_METHOD(Surface::supportsMimeType) {
   auto surface = Nan::ObjectWrap::Unwrap<Surface>(self)->_data;
 
   // in-arguments
-  auto mime_type = *Nan::Utf8String (info[0].As<String>());
+  auto mime_type__value = info[0].As<String>(); auto mime_type = *Nan::Utf8String(mime_type__value);
 
   // function call
   cairo_bool_t result = cairo_surface_supports_mime_type (surface, mime_type);
@@ -630,7 +630,7 @@ NAN_METHOD(Surface::unmapImage) {
 }
 NAN_METHOD(ImageSurface::createFromPng) {
   // in-arguments
-  auto filename = *Nan::Utf8String (info[0].As<String>());
+  auto filename__value = info[0].As<String>(); auto filename = *Nan::Utf8String(filename__value);
 
   // function call
   cairo_surface_t * result = cairo_image_surface_create_from_png (filename);

